### PR TITLE
Implement `class` features in Lox

### PR DESCRIPTION
**Pull request by ryan-gang**
Enhance the Lox interpreter with variable resolution, scoping rules, and class support. Add comprehensive tests for these features, including error handling for semantic issues and variable shadowing. Refactor existing code for improved clarity and maintainability.

## Summary by CodeRabbit

- **New Features**
	- Enhanced object-oriented capabilities allowing users to define classes with methods, properties, initializers, and proper “this” handling.
	- Introduced new classes and methods, including `Spaceship`, `Robot`, `Wizard`, `Superhero`, `Calculator`, and more, with various functionalities.
	- Added functionality for class instantiation within functions and block scopes, supporting visibility and scope rules.
	- Implemented error handling for invalid operations related to class usage and property access.
	- Updated identifiers for test stages to improve clarity and organization.

- **Tests**
	- Expanded test suite covering various class behaviors including instantiation, property access, method invocation, and error scenarios.
	- Added new test cases for class-related functionalities and edge cases, including comprehensive tests for class declarations and method definitions.
	- Enhanced documentation within test cases for better understanding of expected behaviors.

- **Chores**
	- Updated build and test configurations to incorporate the new class functionality and its comprehensive tests.

### CODE REVIEW

**Comment by ryan-gang** [comment]
TODO: take a look at the review for `ast.go`

**Comment by ryan-gang** [comment]
<details><summary>Details</summary>
<p>
Detailed details.
</p>
</details>

File: Makefile [hunk 1]
```diff
@@ -162,4 +162,19 @@ test_resolving_w_jlox: build
...
 
+test_classes_w_jlox: build
+	CODECRAFTERS_REPOSITORY_DIR=./craftinginterpreters/build/gen/chap12_classes \
+	CODECRAFTERS_TEST_CASES_JSON="[ \
+		{\"slug\":\"cl1\",\"tester_log_prefix\":\"stage_801\",\"title\":\"Stage #801: Class Declarations\"}, \
+		{\"slug\":\"cl2\",\"tester_log_prefix\":\"stage_802\",\"title\":\"Stage #802: Class Instances\"}, \
+		{\"slug\":\"cl3\",\"tester_log_prefix\":\"stage_803\",\"title\":\"Stage #803: Getters & Setters\"}, \
+		{\"slug\":\"cl4\",\"tester_log_prefix\":\"stage_804\",\"title\":\"Stage #804: Class & Instance Methods\"}, \
+		{\"slug\":\"cl5\",\"tester_log_prefix\":\"stage_805\",\"title\":\"Stage #805: Bound method calls\"}, \
+		{\"slug\":\"cl6\",\"tester_log_prefix\":\"stage_806\",\"title\":\"Stage #806: this keyword\"}, \
```
Comment (rohitpaulk) [inline Makefile:173]:
```suggestion
		{\"slug\":\"cl6\",\"tester_log_prefix\":\"stage_806\",\"title\":\"Stage #806: The this keyword\"}, \
```

File: test_programs/cl1/2.lox [hunk 2]
```diff
@@ -0,0 +1,9 @@
+---
+expected_error_type: runtime
+---
+{
+  class Dinosaur {}
+  print "Inside block: Dinosaur exists";
```
Comment (rohitpaulk) [inline test_programs/cl1/2.lox:6]: We aren't really testing this, since we aren't doing `print Dinosaur`

File: test_programs/cl1/3.lox [hunk 3]
```diff
@@ -0,0 +1,9 @@
+---
+expected_error_type: none
+---
+// Multiple class declarations
+class Robot {}
```
Comment (rohitpaulk) [inline test_programs/cl1/3.lox:5]: I'd place this as `2.lox` (before the scope testing)

File: test_programs/cl1/4.lox [hunk 4]
```diff
@@ -0,0 +1,11 @@
+---
+expected_error_type: none
+---
+// Class declaration inside function
+fun foo() {
+  class Superhero {}
+  print "Class declared inside function";
```
Comment (rohitpaulk) [inline test_programs/cl1/4.lox:7]: Aren't really testing this, needs to be printed

File: test_programs/cl2/1.lox [hunk 5]
```diff
@@ -0,0 +1,13 @@
+---
+expected_error_type: none
+---
+// Basic class instantiation with conditional
+class Spaceship {}
+var falcon = Spaceship();
+
+if (falcon) {
```
Comment (rohitpaulk) [inline test_programs/cl2/1.lox:8]: Complicated for ths first test - just to `print falcon`

File: test_programs/cl2/2.lox [hunk 6]
```diff
@@ -0,0 +1,15 @@
+---
+expected_error_type: none
+---
+// Multiple instances with a counter
```
Comment (rohitpaulk) [inline test_programs/cl2/2.lox:4]: Let's remove counter, it doesn't test anything specific to the current stage - can be placed in the 4th test if needed, to test unrelated interactions

File: test_programs/cl2/3.lox [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+---
+// Instantiating classes in a function
+class Wizard {}
+class Dragon {}
+
+fun createCharacters() {
+  var merlin = Wizard();
+  var smaug = Dragon();
+  print "Characters created in fantasy world:";
+  print merlin;
```
Comment (rohitpaulk) [inline test_programs/cl2/3.lox:12]: Either this (test 3) or test2 would be good place to test truthiness of instances - currently it is tested in test 1

File: test_programs/cl3/4.lox [hunk 8]
```diff
@@ -0,0 +1,16 @@
+---
+expected_error_type: none
+---
+// Multiple instances with properties
+class Superhero {}
```
Comment (rohitpaulk) [inline test_programs/cl3/4.lox:5]: 3 and 4 could probably be switched since 3 is more of a combination test

File: test_programs/cl4/1.lox [hunk 9]
```diff
@@ -0,0 +1,20 @@
+---
+expected_error_type: none
+---
+class Robot {
+  beep() {
```
Comment (rohitpaulk) [inline test_programs/cl4/1.lox:5]: Do one test that only has methods without arguments, and then include arguments from test 2 onwards

File: test_programs/cl4/3.lox [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+class Dragon {
+  breatheFire() {
+    print "Breathing fire!";
+  }
+}
+
+// Methods on different class instances
+var merlin = Wizard();
+var smaug = Dragon();
```
Comment (rohitpaulk) [inline test_programs/cl4/3.lox:18]: We need to test multiple arguments more

File: test_programs/cl5/1.lox [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+dog.sound = "Woof";
+dog.species = "Dog";
+
+var cat = Animal();
+cat.sound = "Meow";
+cat.species = "Cat";
+
+// Swap methods between instances
+cat.makeSound = dog.makeSound;
+dog.identify = cat.identify;
```
Comment (rohitpaulk) [inline test_programs/cl5/1.lox:24]: TODO: Think about whether this needs to be a differnt stage, or merged into previous. 

New things that we're testing here: 

- Accessing a method as a value (a.b, without a.b())
- Whether methods are bound to their instance

File: test_programs/cl6/2.lox [hunk 12]
```diff
@@ -0,0 +1,14 @@
+---
+expected_error_type: none
+---
+class Cake {
+  taste() {
+    var adjective = "delicious";
+    print "The " + this.flavor + " cake is " +
+    adjective + "!";
```
Comment (rohitpaulk) [inline test_programs/cl6/2.lox:8]: Let's not test this here, keep it simple to understand what the test expects

File: test_programs/cl6/3.lox [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+      print "Casting spell as " + this.name;
+    }
+
+    return castSpell;
+  }
+}
+
+var wizard = Wizard();
+wizard.name = "Merlin";
+var spellCaster = wizard.getSpellCaster();
```
Comment (rohitpaulk) [inline test_programs/cl6/3.lox:17]: Could also testL

```suggestion
wizard.getSpellCaster()();
```

File: test_programs/cl5/1.lox (hunk 11 above)
Comment (rohitpaulk) [reply test_programs/cl5/1.lox:24]: We should merge this into the next stage (the "The this keyword" stage). It can be test-2 or test-3.

File: Makefile [hunk 14]
```diff
@@ -162,4 +162,19 @@ test_resolving_w_jlox: build
...
+		{\"slug\":\"cl7\",\"tester_log_prefix\":\"stage_807\",\"title\":\"Stage #807: this keyword errors\"}, \
```
Comment (rohitpaulk) [inline Makefile:174]:
```suggestion
		{\"slug\":\"cl7\",\"tester_log_prefix\":\"stage_807\",\"title\":\"Stage #807: Invalid usages of this\"}, \
```

File: test_programs/cl7/4.lox [hunk 15]
```diff
@@ -0,0 +1,14 @@
+---
+expected_error_type: none
+---
+class Confused {
+  method() {
+    fun inner(instance) {
```
[the remark this answers is not in the document]
Comment (rohitpaulk) [reply test_programs/cl7/4.lox:6]: We could even just remove this, since for errors simple tests are sufficient

File: test_programs/cl8/3.lox [hunk 16]
```diff
@@ -0,0 +1,24 @@
+---
+expected_error_type: none
+---
+class Counter {
+  init(startValue) {
+    if (startValue < 0) {
+      print "startValue can't be negative";
```
Comment (rohitpaulk) [inline test_programs/cl8/3.lox:7]: test 1 & 2 here look okay - they test the constructor behaviour specifically and aren't too complex. test 4 is okay too, tests a combination. test 3 here could instead test calling init on an instance

File: test_programs/cl9/1.lox [hunk 17]
```diff
@@ -0,0 +1,16 @@
+---
+expected_error_type: none
+---
+class Person {
+  init() {
+    this.name = "Ryan";
+    this.age = 25;
+    return;
```
Comment (rohitpaulk) [inline test_programs/cl9/1.lox:8]:
```suggestion
    return this;
```

File: test_programs/cl9/1.lox [hunk 18]
```diff
@@ -0,0 +1,16 @@
+---
+expected_error_type: none
+---
+class Person {
+  init() {
+    this.name = "Ryan";
```
Comment (rohitpaulk) [inline test_programs/cl9/1.lox:6]: Remove all of this, just say `print "init()"`

File: test_programs/cl9/1.lox [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+    this.age = 25;
+    return;
+  }
+
+  sayHello() {
```
Comment (rohitpaulk) [inline test_programs/cl9/1.lox:11]: Also remove this, distracting

File: Makefile [hunk 20]
```diff
@@ -162,4 +162,19 @@ test_resolving_w_jlox: build
...
+		{\"slug\":\"cl8\",\"tester_log_prefix\":\"stage_808\",\"title\":\"Stage #808: Constructor calls\"}, \
+		{\"slug\":\"cl9\",\"tester_log_prefix\":\"stage_809\",\"title\":\"Stage #809: Constructor errors\"} \
```
Comment (rohitpaulk) [inline Makefile:176]:
```suggestion
		{\"slug\":\"cl9\",\"tester_log_prefix\":\"stage_809\",\"title\":\"Stage #809: Return within constructors\"} \
```
